contracts/TreasuryInsurance.sol

```diff
@@ -0,0 +1,100 @@
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts/token/ERC721/ERC721.sol';
+import './interfaces/IERC20Allowance.sol';
+import './Treasury.sol';
+
+// Treasury contract with insurance.
+contract TreasuryInsurance is Treasury, ERC721 {
+    mapping(uint256 => uint256) public compensationLimits;
+    mapping(uint256 => string) public insuranceConditions;
+    mapping(uint256 => uint256) public insuranceCosts;
+    mapping(uint256 => address) public insuranceAssets;
+    mapping(uint256 => uint256) public lastPayTime;
+    mapping(uint256 => uint256) public requestedCompensations;
+    mapping(uint256 => string) public requestedCases;
+
+    uint256 public insuranceCounter;
+
+    event Insured(
+        uint256 id,
+        address insured,
+        address asset,
+        uint256 cost,
+        uint256 compensationLimit
+    );
+    event Requested(uint256 id, uint256 compensation);
+    event PaidInsurance(uint256 id, uint256 payment);
+    event Compensated(uint256 id, address insured, uint256 compensationLimit);
+
+    constructor(address governance_, address executor_)
+        Treasury(governance_, executor_)
+        ERC721('BITDAO_TREASURY_INSURANCE', 'BITTI')
+    {}
+
+    function insure(
+        address to,
+        address asset,
+        uint256 cost,
+        uint256 compensationLimit,
+        string memory condition
+    ) external virtual onlyExecutor returns (uint256) {
+        compensationLimits[insuranceCounter] = compensationLimit;
+        insuranceConditions[insuranceCounter] = condition;
+        insuranceCosts[insuranceCounter] = cost;
+        insuranceAssets[insuranceCounter] = asset;
+        lastPayTime[insuranceCounter] = block.timestamp;
+        _safeMint(to, insuranceCounter);
+        insuranceCounter += 1;
+        emit Insured(insuranceCounter, to, asset, cost, compensationLimit);
+        return insuranceCounter;
+    }
+
+    function payInsurance(uint256 id, uint256 period) public virtual {
+        lastPayTime[id] = block.timestamp + period;
+        IERC20(insuranceAssets[id]).transferFrom(
+            _msgSender(),
+            address(this),
+            period * insuranceCosts[id]
+        );
+        emit PaidInsurance(id, period * insuranceCosts[id]);
+    }
+
+    function requestInsurance(
+        uint256 id,
+        uint256 compensation,
+        string memory reason
+    ) external virtual {
+        require(
+            ownerOf(id) == _msgSender(),
+            'Insurance::request: requester must be owner'
+        );
+        require(
+            compensation <= compensationLimits[id],
+            'Insurance::request: requested compensation is more than the limit'
+        );
+        requestedCompensations[id] = compensation;
+        requestedCases[id] = reason;
+        if (block.timestamp > lastPayTime[id]) {
+            payInsurance(id, block.timestamp - lastPayTime[id]);
+        }
+        emit Requested(id, compensation);
+    }
+
```

Comment:
To me the below "compensate()" function missed invoking a payout function to the user after we set his requestedCompensation[id] = 0.